dbms/src/Interpreters/NullAwareSemiJoinHelper.h

```diff
@@ -0,0 +1,157 @@
+// Copyright 2023 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <Common/Logger.h>
+#include <Core/Block.h>
+#include <Interpreters/Join.h>
+#include <Parsers/ASTTablesInSelectQuery.h>
+
+
+namespace DB
+{
+
+enum class NASemiJoinStep : UInt8
+{
+    /// Check other conditions for the right rows whose join key are equal to this left row.
+    /// The join keys of this left row must not have null.
+    NOT_NULL_KEY_CHECK_MATCHED_ROWS,
+    /// Check join key equal condition and other conditions(if any) for the right rows
+    /// with at least one null join key.
+    /// The join keys of this left row must not have null.
+    NOT_NULL_KEY_CHECK_NULL_ROWS,
+    /// Like `CHECK_NULL_ROWS_NOT_NULL` except the join keys of this left row must have null.
+    NULL_KEY_CHECK_NULL_ROWS,
+    /// Check join key equal condition and other conditions(if any) for all right rows in blocks.
+    /// The join keys of this left row must have null.
+    NULL_KEY_CHECK_ALL_BLOCKS,
+    /// Work is done.
+    DONE,
+};
+
+enum class NASemiJoinResultType : UInt8
+{
+    FALSE_VALUE,
+    TRUE_VALUE,
+    NULL_VALUE,
+};
```

Comment:
Although `MATCHED` may be a more straightforward name, I think the current result does not add more complexity because it's fairly simple just to consider the result of semi-join when calling `setResult`.

Reply:
but we actually have three join types